### PR TITLE
Ensure all autologging callbacks are picklable

### PR DESCRIPTION
**Pull request by harupy**
Signed-off-by: harupy <hkawamura0130@gmail.com>

## What changes are proposed in this pull request?

Ensure all autologging callbacks are picklable.

## How is this patch tested?

Unit tests

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: tests/xgboost/test_xgboost_autolog.py [hunk 1]
```diff
@@ -554,3 +555,20 @@ def test_xgb_autolog_does_not_break_dmatrix_instantiation_with_data_none():
...
+        xgb.train(params, dtrain, evals=[(dtrain, "train")])
+    return run.info.run_id
+
+
+def test_xgb_autolog_multi_processing(bst_params, tmpdir):
+    tracking_uri = tmpdir.join("mlruns").strpath
+    with mp.Pool(2) as pool:
+        run_ids = pool.starmap(train, [(tracking_uri, bst_params)] * 2)
+    client = mlflow.tracking.MlflowClient(tracking_uri)
+    assert all(client.get_run(run_id).data.params.keys() >= bst_params.keys() for run_id in run_ids)
```
Comment (harupy) [inline tests/xgboost/test_xgboost_autolog.py:574]: I'm adding tests to reveal autologging integrations that don't work with `multiprocessing`.

**Comment by harupy** [comment]
Interesting. Lots of tests hung.

File: mlflow/ml-package-versions.yml [hunk 2]
```diff
@@ -77,7 +77,8 @@ tensorflow:
       ">= 2.6.0, < 2.7.0": ["tensorflow-estimator < 2.7.0"]
       "== dev": ["scikit-learn"]
     run: |
-      pytest tests/tensorflow/test_tensorflow2_autolog.py --large
+      pytest tests/tensorflow/test_tensorflow2_autolog.py --large -k 'not multiprocessing'
+      pytest tests/tensorflow/test_tensorflow2_autolog.py --large -k multiprocessing
```
Comment (WeichenXu123) [inline mlflow/ml-package-versions.yml:81]: why need split it into 2 pytest commands ?

Comment (harupy) [reply mlflow/ml-package-versions.yml:81]: I'll insert a comment for this, but when we run a multiprocessing test with other tests, it hangs.

**Comment by harupy** [comment]
Fastai doesn't seem to support distributed training using CPUs:

https://docs.fast.ai/distributed.html

**Comment by harupy** [comment]
Distributed training in MXnet seems pretty complex:

https://mxnet.apache.org/versions/1.8.0/api/faq/distributed_training

**Comment by harupy** [comment]
For sklearn, we already have a test for parallelised training:

https://github.com/mlflow/mlflow/blob/d6ae84164987e2881cdffdce07f7dcf0e3d94aaa/tests/sklearn/test_sklearn_autolog.py#L732-L733

Is there anything else that we need to test for scikit-learn?

**Review by dbczumar — COMMENTED** [review]
@harupy Looks great! Can we address LightGBM as well?

**Comment by dbczumar** [comment]
> For sklearn, we already have a test for parallelised training:
> 
> https://github.com/mlflow/mlflow/blob/d6ae84164987e2881cdffdce07f7dcf0e3d94aaa/tests/sklearn/test_sklearn_autolog.py#L732-L733
> 
> Is there anything else that we need to test for scikit-learn?

Does this test case use a multiprocess backend?

**Comment by harupy** [comment]
@dbczumar Thanks for the comment!

> Does this test case use a multiprocess backend?

We run the test using `lowky` which is a multi-processing based backend:

https://github.com/mlflow/mlflow/blob/d6ae84164987e2881cdffdce07f7dcf0e3d94aaa/tests/sklearn/test_sklearn_autolog.py#L715-L733

**Comment by dbczumar** [comment]
> @dbczumar Thanks for the comment!
> 
> > Does this test case use a multiprocess backend?
> 
> We run the test using `lowky` which is a multi-processing based backend ([doc](https://scikit-learn.org/stable/computing/parallelism.html#parallelism)):
> 
> https://github.com/mlflow/mlflow/blob/d6ae84164987e2881cdffdce07f7dcf0e3d94aaa/tests/sklearn/test_sklearn_autolog.py#L715-L733

Sounds good!

**Comment by harupy** [comment]
@dbczumar 

> Can we address LightGBM as well?

Done!

File: mlflow/utils/autologging_utils/safety.py [hunk 3]
```diff
@@ -51,6 +51,27 @@ def safe_function(*args, **kwargs):
...
+
+def picklable_exception_safe_function(function):
+    """
+    Wraps the specified function with broad exception handling to guard
+    against unexpected errors during autologging while preserving picklability.
+    """
+    if is_testing():
+        setattr(function, _ATTRIBUTE_EXCEPTION_SAFE, True)
+
+    return update_wrapper_extended(functools.partial(_safe_function, function), function)
```
Comment (harupy) [inline mlflow/utils/autologging_utils/safety.py:72]: I initially tried this:

```diff
diff --git a/mlflow/utils/autologging_utils/safety.py b/mlflow/utils/autologging_utils/safety.py
index 11ffde645..4b15e38ea 100644
--- a/mlflow/utils/autologging_utils/safety.py
+++ b/mlflow/utils/autologging_utils/safety.py
@@ -96,7 +96,7 @@ def _exception_safe_class_factory(base_class):
     class _ExceptionSafeClass(base_class):
         def __new__(cls, name, bases, dct):
             for m in dct:
                 # class methods or static methods are not callable.
                 if callable(dct[m]):
-                    dct[m] = exception_safe_function(dct[m])
+                    dct[m] = picklable_exception_safe_function(dct[m])
             return base_class.__new__(cls, name, bases, dct)
 
     return _ExceptionSafeClass
```

but this gave:

```python
% pytest tests/xgboost/test_xgboost_autolog.py -k is_pickable
============================================================ test session starts =============================================================
platform linux -- Python 3.7.10, pytest-6.2.5, py-1.10.0, pluggy-1.0.0 -- /home/haru/miniconda3/envs/mlflow-dev-env/bin/python
cachedir: .pytest_cache
rootdir: /home/haru/Desktop/repositories/mlflow, configfile: pytest.ini
collected 29 items / 27 deselected / 2 selected                                                                                              

tests/xgboost/test_xgboost_autolog.py::test_callback_func_is_pickable PASSED                                                           [ 50%]
tests/xgboost/test_xgboost_autolog.py::test_callback_class_is_pickable FAILED                                                          [100%]

================================================================== FAILURES ==================================================================
______________________________________________________ test_callback_class_is_pickable _______________________________________________________

    @pytest.mark.skipif(
        Version(xgb.__version__.replace("SNAPSHOT", "dev")) < Version("1.3.0"),
        reason="`xgboost.callback.TrainingCallback` is not supported",
    )
    def test_callback_class_is_pickable():
        from mlflow.xgboost._autolog import AutologCallback
    
>       cb = AutologCallback(BatchMetricsLogger(run_id="1234"), eval_results={})

AutologCallback = <class 'mlflow.xgboost._autolog.AutologCallback'>

tests/xgboost/test_xgboost_autolog.py:577: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

function = <function AutologCallback.__init__ at 0x7fd1c2699dd0>
args = (<mlflow.utils.autologging_utils.BatchMetricsLogger object at 0x7fd19e0ac890>,), kwargs = {'eval_results': {}}

    def _safe_function(function, *args, **kwargs):
        try:
>           return function(*args, **kwargs)
E           TypeError: __init__() missing 1 required positional argument: 'metrics_logger'

args       = (<mlflow.utils.autologging_utils.BatchMetricsLogger object at 0x7fd19e0ac890>,)
function   = <function AutologCallback.__init__ at 0x7fd1c2699dd0>
kwargs     = {'eval_results': {}}

mlflow/utils/autologging_utils/safety.py:56: TypeError
```

Comment (dbczumar) [reply mlflow/utils/autologging_utils/safety.py:72]: @harupy This solution makes sense once we can resolve the `__init__` issues. Let's also make sure to remove `exception_safe_function` and replace it with `picklable_exception_safe_function`.

Comment (harupy) [reply mlflow/utils/autologging_utils/safety.py:72]: This [SO post](https://stackoverflow.com/a/16626797) says:

> partial objects are like function objects in that they are callable, weak referencable, and can have attributes. There are some important differences. For instance, the __name__ and __doc__ attributes are not created automatically. Also, partial objects defined in classes behave like static methods and **do not transform into bound methods during instance attribute look-up**.

Comment (dbczumar) [reply mlflow/utils/autologging_utils/safety.py:72]: Got it. Let's ignore https://github.com/mlflow/mlflow/pull/5039/files#r750767771 and use a separate function for methods defined on classes :). Thanks @harupy !

**Review by dbczumar — APPROVED** [review]
LGTM! Thanks a bunch, @harupy!

**Pull request merged**